### PR TITLE
update conan v2 imports for main recipe

### PR DESCRIPTION
**Pull request by paulocoutinhox**

### CODE REVIEW

File: conan/recipe/conanfile.py [hunk 1]
```diff
@@ -45,59 +47,60 @@ class TargetConan(ConanFile):
         "nativium_code_coverage": False,
     }
     exports_sources = "*"
-    generators = "cmake"
+    generators = "CMakeToolchain", "CMakeDeps"
```
Comment (prince-chrismc) [inline conan/recipe/conanfile.py:50]: Having a generator in two place has weird behavior,
Toolchain is in `generare()` so you dont want it here in the attribute

```suggestion
    generators = "CMakeDeps"
```

Comment (paulocoutinhox) [reply conan/recipe/conanfile.py:50]: Hi @prince-chrismc. Thanks for this. But after apply, i run again it does nothing different. The problem is same.

**Comment by prince-chrismc** [comment]
Hmm the install is being ran from the log https://github.com/nativium/nativium/blob/5425ca3f53fe506076dcbfaf1ff3fd954235c3d9/targets/linux/verbs/setup.py#L46 can you confirm I understood?


🤔 your layout says the source and build are the current directory. But I dont see the cmake list there.

I think you might have missed pointing the layout to the correct locations

**Comment by prince-chrismc** [comment]
https://docs.conan.io/en/latest/developing_packages/package_layout.html#declaring-the-layout

You root seems to be 2 or 3 levels up.

https://github.com/prince-chrismc/user-management/blob/5578bdef0de6e4e44265a021faa48bb87a855570/backend/conanfile.py#L20

You CMAKELISTS.TXT is in the root so you do not need the subfolder AFAIK

**Comment by paulocoutinhox** [comment]
Hi @prince-chrismc,

I tried this:

```
# -----------------------------------------------------------------------------
def layout(self):
    self.folders.root = os.path.join("..", "..")
    cmake_layout(self)
```

But in my case the command is executed from root:

```
python3 nativium.py target linux setup
```

It will run conan install inside build folder:

```
run_args = [
    "conan",
    "install",
    os.path.join(
        proj_path,
        "conan",
        "recipe",
        const.FILE_NAME_CONANFILE_PY,
    ),
    "-pr:b",
    build_profile,
    "-pr:h",
    os.path.join(
        proj_path, "conan", "profiles", arch["conan_profile"]
    ),
    "-s:h",
    "os.version={0}".format(arch["min_version"]),
]

target.add_target_setup_common_args(
    run_args, target_name, target_config, arch, build_type
)

run_args.append("--build=missing")
run_args.append("--update")
```

```
build_dir = os.path.join(
    proj_path,
    "build",
    target_name,
    build_type,
    arch["arch"],
    "conan",
)
```

It will install into "build/[...]/conan" folder without the cmake toolchain files for build.

And after, need run "build" target verb, that will build with my CMakeLists.txt and need pass/import the cmake file generated by setup command (this flow is correct for v2?):

```
python3 nativium.py target linux build
```

It is strange, because with a new project (`conan new hello/0.1 -m=cmake_lib`), it works and generate the correct files when i run `conan install .`:

image

**Comment by prince-chrismc** [comment]
You should read the layout docs I linked above. The intro explains what it's trying to do... were pretty swamped so I am not sure I'll have time to checkout and play with this but I'll try to answer you questions.

***

You need to tell the recipe where to get information and where to put files. That's the role of layout.
The structure is really unique for your project. So the default is not enough.

I would suggest manually setting up the layout. Set all the properties the same way you building framework does.

I am not sure how you are using the template from `conan new` but if you more the files you'll see things break.

Please push the code so I can see the logs 🙏 or post the error message.

**Comment by paulocoutinhox** [comment]
Hi @prince-chrismc, i have a lot of questions rsrsrsrs

In my case, i only use conan for dependencies and not for build my full project. The `conanfile.py` is only to generate the `.cmake` to be imported into my CMakeLists.txt because of dependencies. And at least i link with all conan installed libs:

```
target_link_libraries(${NATIVIUM_PROJECT_NAME} "${CONAN_LIBS}")
```

Current questions:

1 - Current project generate file `conanbuildinfo.cmake` that i include in my cmake (using `generators = "cmake"`). This is not possible more?
2 - If number 1 is NO, what is correct way to use CMake?

**Comment by prince-chrismc** [comment]
That's a fair point! 

https://github.com/conan-io/conan-io.github.io/pull/222 I call out a few things in that blog

The new `CMakeToolchain` generator use the cmake toolchain mechanism 😆 https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html that's how cmake knows to use the config files from Conan and a few other important settings 

You should remove everything Conan specific from your cmake and use modern cmake to link your dependencies

**Comment by paulocoutinhox** [comment]
@prince-chrismc i call you on google chat...here is too slow... or we can talk on slack

I defined my layout vars:

```
def layout(self):
    generators_dir = os.path.join(
        "build",
        self.options.get_safe("nativium_target").value,
        self.options.get_safe("nativium_build_type").value,
        self.options.get_safe("nativium_arch").value,
        "conan",
    )

    build_dir = os.path.join(
        "build",
        self.options.get_safe("nativium_target").value,
        self.options.get_safe("nativium_build_type").value,
        self.options.get_safe("nativium_arch").value,
        "target"
    )

    self.folders.root = os.path.join("..", "..")
    self.folders.source = "."
    self.folders.build = build_dir
    self.folders.generators = generators_dir

    cmake_layout(self)
```

1 - When i call conan install the `generators` are generated in `build/generators` instead of my `generators` folder above.

image

2 - A file called `CMakeUserPresets.json` is always generated on root, it is necessary?

3 - The folder "source", "build", "generators" is appended to "root" folder, correct?

**Comment by prince-chrismc** [comment]
I'll slack you an appointment slot in the AM :) we can do a zoom hopefully before I leave 

My night is your morning but should work nicely

**Comment by paulocoutinhox** [comment]
My english speak is bad, im BR ... i think that is better something to write rsrsrsrsrsrs There is any option?

Im "Paulo Coutinho" on slack "Cpplang".

**Comment by paulocoutinhox** [comment]
@prince-chrismc i discover the problem after debug conan code.

I set all the folder but when i call "cmake_layout(self)", it change all the folder. Example:

```
def cmake_layout(conanfile, generator=None, src_folder=".", build_folder="build"):
    [...]
    build_folder = build_folder if not subproject else os.path.join(subproject, build_folder)
    custom_conf = get_build_folder_custom_vars(conanfile)
    if custom_conf:
        build_folder = os.path.join(build_folder, custom_conf)

    if multi:
        conanfile.folders.build = build_folder
    else:
        conanfile.folders.build = os.path.join(build_folder, build_type)

    conanfile.folders.generators = os.path.join(build_folder, "generators")
    [...]
```

Basically, i cant change it.

**Comment by paulocoutinhox** [comment]
@prince-chrismc I created a project to test and understand the details and it appear to work now:
https://github.com/paulocoutinhox/conan-cmake-v2

**Pull request merged**